runtime/src/iree/hal/drivers/cuda/cuda_device.h

```diff
@@ -23,6 +24,14 @@ iree_status_t iree_hal_cuda_device_create(
     iree_hal_cuda_dynamic_symbols_t* syms, CUdevice device,
     iree_allocator_t host_allocator, iree_hal_device_t** out_device);
 
+// Returns `true` if `base_device` is an instance of a HAL CUDA device.
+bool iree_hal_is_cuda_device(iree_hal_device_t* base_device);
+
+// Returns a pointer to context wrapper owned by the `base_device`. Undefined
+// if `base_device` is not a HAL CUDA device.
+iree_hal_cuda_context_wrapper_t* iree_hal_cuda_device_context_wrapper(
```

Comment:
do you need this to be public? The context wrapper is some internal data structure. How would this be used?

Reply:
Yes, in https://github.com/iree-org/iree-samples/pull/126 I need to detect if device is a CUDA device, and get access to the context (and maybe to the CUStream, but that will come later).

In theory I can use primary context, and looks like I don't need CUDevice (at least now), but what if IREE will use non-primary context?

Reply:
cudnnCreateHandle has to be called with the correct context, and maybe some other cuDNN APIs, and given that we can run on a system with multiple GPUs, looks like it has to be available through CUDA hal device. Though in some sense openxla-nvgpu is a "privileged plugin", I think it will need to be aware of a lot of CUDA HAL internals to get peak performance

Reply:
Would it make more sense to have a function that query the CUcontext then? Instead of getting the wrapper?

Reply:
Done. Replaced with a single `iree_hal_cuda_device_get_context` function, CUdevice can be resolved from CUcontext.